pkg/controllers/oauthendpoints/oauth_endpoints_controller.go

```diff
@@ -0,0 +1,263 @@
+		return listOAuthRoutes(routeLister)
+	}
+
+	getTLSConfigFunc := func() (*tls.Config, error) {
+		return getOAuthRouteTLSConfig(cmLister, secretLister, ingressLister, systemCABundle)
+	}
+
+	return endpointaccessible.NewEndpointAccessibleController(
+		"OAuthServerRoute",
+		operatorClient,
+		endpointListFunc, getTLSConfigFunc,
+		[]factory.Informer{
+			cmInformer,
+			secretInformer,
+			routeInformer,
+			ingressInformer,
+		},
+		recorder)
+}
+
+// NewOAuthServiceCheckController returns a controller that checks the health of authentication service.
+func NewOAuthServiceCheckController(
+	operatorClient v1helpers.OperatorClient,
+	kubeInformersForTargetNS informers.SharedInformerFactory,
+	recorder events.Recorder,
+) factory.Controller {
+	endpointsListFunc := func() ([]string, error) {
+		return listOAuthServices(kubeInformersForTargetNS.Core().V1().Services().Lister())
+	}
+
+	getTLSConfigFunc := func() (*tls.Config, error) {
+		return getOAuthEndpointTLSConfig(kubeInformersForTargetNS.Core().V1().ConfigMaps().Lister())
+	}
+
+	return endpointaccessible.NewEndpointAccessibleController(
+		"OAuthServerService",
+		operatorClient,
+		endpointsListFunc, getTLSConfigFunc,
+		[]factory.Informer{
+			kubeInformersForTargetNS.Core().V1().ConfigMaps().Informer(),
+			kubeInformersForTargetNS.Core().V1().Services().Informer(),
+		},
+		recorder)
+}
+
+// NewOAuthServiceEndpointsCheckController returns a controller that checks the health of authentication service
+// endpoints.
+func NewOAuthServiceEndpointsCheckController(
+	operatorClient v1helpers.OperatorClient,
+	kubeInformersForTargetNS informers.SharedInformerFactory,
+	recorder events.Recorder,
+) factory.Controller {
+	endpointsListFn := func() ([]string, error) {
+		return listOAuthServiceEndpoints(kubeInformersForTargetNS.Core().V1().Endpoints().Lister())
+	}
+
+	getTLSConfigFunc := func() (*tls.Config, error) {
+		return getOAuthEndpointTLSConfig(kubeInformersForTargetNS.Core().V1().ConfigMaps().Lister())
+	}
+
+	return endpointaccessible.NewEndpointAccessibleController(
+		"OAuthServerServiceEndpoints",
+		operatorClient,
+		endpointsListFn, getTLSConfigFunc,
+		[]factory.Informer{
+			kubeInformersForTargetNS.Core().V1().Endpoints().Informer(),
+			kubeInformersForTargetNS.Core().V1().ConfigMaps().Informer(),
+		},
+		recorder)
+}
+
+func listOAuthServiceEndpoints(endpointsLister corev1listers.EndpointsLister) ([]string, error) {
+	var results []string
+	endpoints, err := endpointsLister.Endpoints("openshift-authentication").Get("oauth-openshift")
+	if err != nil {
+		return nil, fmt.Errorf("failed to get oauth service endpoints: %w", err)
+	}
+	for _, subset := range endpoints.Subsets {
+		for _, address := range subset.Addresses {
+			for _, port := range subset.Ports {
+				results = append(results, net.JoinHostPort(address.IP, strconv.Itoa(int(port.Port))))
+			}
+		}
+	}
+	if len(results) == 0 {
+		return nil, fmt.Errorf("oauth service endpoints are not ready")
```

Comment:
this will be `degraded` not `progressing`, is that okay?

Reply:
if yes, then we could update the msg

Reply:
I think I'm currently fine with both the message and the status change, we can see if it actually ever becomes an issue and fix it later if it is